website/content/docs/upgrade/upgrade-specific.mdx

```diff
@@ -980,24 +989,41 @@ as shown in commands like `nomad server members`. To see the version of the Raft
 protocol in use on each server, use the `nomad operator raft list-peers`
 command.
 
-The easiest way to upgrade servers is to have each server leave the cluster,
-upgrade its `raft_protocol` version in the `server` stanza, and then add it
-back. Make sure the new server joins successfully and that the cluster is stable
-before rolling the upgrade forward to the next server. It's also possible to
-stand up a new set of servers, and then slowly stand down each of the older
-servers in a similar fashion.
-
 When using Raft protocol version 3, servers are identified by their `node-id`
 instead of their IP address when Nomad makes changes to its internal Raft quorum
 configuration. This means that once a cluster has been upgraded with servers all
 running Raft protocol version 3, it will no longer allow servers running any
-older Raft protocol versions to be added. If running a single Nomad server,
-restarting it in-place will result in that server not being able to elect itself
-as a leader. To avoid this, either set the Raft protocol back to 2, or use
-[Manual Recovery Using
+older Raft protocol versions to be added.
+
+~> **Warning:** If you are running a single Nomad server, restarting it
+in-place will result in that server not being able to elect itself as
+a leader. To avoid this, either set the Raft protocol back to 2, or
+use [Manual Recovery Using
```

Comment:
Hm so there's no way to upgrade a single server from 2 -> 3 inplace? A second server or `peers.json` _must_ be used?

If so I think we should at least make an issue to track that and document exactly how to use peers.json to fix it. Seems like if a user can fix it by manually updating peers.json we may be able to at least automate that or otherwise special case when there's only a single server member.

Reply:
Alternatively we could make `/docs/upgrade/upgrade-specific#upgrading-to-raft-protocol-3` its own page with details for Standalone and HA clusters.

Reply:
Yeah, the terrible thing is that I just looked at the Outage Recovery tutorial and the first section is ["if your single-server cluster fails, tough luck"](https://learn.hashicorp.com/tutorials/nomad/outage-recovery#failure-of-a-single-server-cluster). 😀 

I'm thinking we can move the raft protocol upgrade section out of these version-specific sections to a top-level section in the upgrade overview page. And that'll include whatever needs to be done to make the peers.json recovery work for a single server.

Reply:
Recorded the workflow for a standalone server in the following script. We'll tell folks to update their config file and then run the equivalent of this for single-node clusters.

```bash
#!/usr/bin/env bash
set -e

err_exit() {
    echo "$1"
    exit 1
}

command -v jq > /dev/null || err_exit "requires jq"
command -v nomad > /dev/null || err_exit "requires nomad"

PEERS=$(nomad agent-info -json | jq -r .stats.raft.num_peers)
if [[ "$PEERS" != "0" ]]; then
    err_exit "this script only works for clusters with a single server"
fi

NOMAD_DATA_DIR=$(nomad agent-info -json | jq -r '.config.DataDir')
NOMAD_ADDR=$(nomad agent-info -json | jq -r '.stats.nomad.leader_addr')
NODE_ID=$(cat "$NOMAD_DATA_DIR/server/node-id")

cat <<EOF > "$NOMAD_DATA_DIR/server/raft/peers.json"
[
  {
    "id": "$NODE_ID",
    "address": "$NOMAD_ADDR",
    "non_voter": false
  }
]
EOF

systemctl restart nomad
```

Reply:
From our sidebar conversation: adding another note here to myself that the `Protocol` field in `nomad server members` has reported 2 since 0.5.0 released in 2014 ([ref](https://github.com/hashicorp/serf/blob/master/CHANGELOG.md#050-march-12-2014)) so I'm going to update this to change to be the raft protocol instead.

Reply:
And one more item from another chat: we want this to deprecate raft v2 in 1.3.0 so that we can remove raft v2 in 1.4.0. That'll make it feasible to remove the current janky Consul import we have to pull in an old version of autopilot and replace it with the proper library version https://github.com/hashicorp/raft-autopilot

Reply:
@schmichael I've updated the docs here as discussed. I haven't touched the `Protocol` field in `nomad server members` because you've got https://github.com/hashicorp/nomad/pull/11600 open and it's probably better to keep that change all in one place.

Reply:
Ha, sorry to muddle things with that. Sometimes you start tracing a thread and can't help but pull it.